[SR-12851] System-wide cache of SwiftPM dependencies

tgymnich opened the pull request:
I reworked my approach from PR #2835 to not only operate on `GitRepository` but on any kind of repository from within `Repository Manager`.

[SR-12851](https://bugs.swift.org/browse/SR-12851)

This PR requires the [Lock Improvements](https://github.com/apple/swift-tools-support-core/pull/139) in swift-tools-support-core to be merged.

neonichu commented:
@swift-ci please smoke test

neonichu commented:
Great, thank you! I have yet to actually review the code, but did try out a SwiftPM with your diff for a bit myself and it seems to be working nicely.

In terms of speedup, I'm seeing a fresh `swift package resolve` of SwiftPM itself go down from ~25s to ~15s when using the cache. Copying from the cache takes up quite a bit of time here, but that's expected since we are going with a very conservative approach.

tgymnich commented:
The tests are also using the local cache. We might need another change to swift-tools-support to modify  `SwiftPMProduct.SwiftPackage.execute(_ args: [String], packagePath: AbsolutePath? = nil, env: [String : String]? = nil))` maybe adding a `--skip-cache` flag by default.

neonichu commented on Sources/SourceControl/RepositoryManager.swift:
```diff
@@ -267,10 +280,69 @@ public class RepositoryManager {
-                        try self.provider.fetch(repository: handle.repository, to: repositoryPath)
+                        let cache = try await { self.cacheManager?.lookup(repository: handle.repository, completion: $0) }
+
+                        try self.provider.fetch(repository: cache.repository, to: repositoryPath)
```
I think we should look into whether a straight up copy might be faster than performing a fetch here.

abertelrud replied:
That's a good idea, and should also eliminate the need to reset the remote to point back to the original URL later.

If that were done in a way that used no-copy cloning on file systems that support it (e.g. APFS), that could be really fast since the bare repository typically has few but large files (typically mainly a single pack).

neonichu commented:
This is great, I have three high-level asks:
- I think we should consider having an opt-out for the cache as a flag, it might be beneficial in some situations to not use the cache at all.
- Could we look into making sure the "Fetching X" log statements reflect whether the fetch is coming from the cache?
- Can we add some tests, especially around concurrent access to the cache? Anecdotally, this seems to be working fine for me locally, but we should have tests to keep that working over time.

neonichu commented:
BTW, the self-hosted tests are still using a vendored copy of TSC. This is something I am working on getting rid off, but we're not quite there, yet. So for now, you would need to update the vendored copy with the lock changes, simply by copying those files over.

neonichu commented:
One more thing, it's great that we are using the cache directory as determined by Foundation for storage, but could we also add a symlink from `~/.swiftpm/repository-cache` to that directory? This makes it easier for users which may have experience with other CLI tools, especially on non-Apple platform, to find the cache directory if they need to.

tomerd commented:
> I think we should consider having an opt-out for the cache as a flag, it might be beneficial in some situations to not use the cache at all.

do we already have a command for clearing the cache? I can see that being useful in case the cache gets corrupt for whatever reason

tgymnich commented:
> BTW, the self-hosted tests are still using a vendored copy of TSC. This is something I am working on getting rid off, but we're not quite there, yet. So for now, you would need to update the vendored copy with the lock changes, simply by copying those files over.

Should I open a separate PR for that?

tgymnich commented:
> One more thing, it's great that we are using the cache directory as determined by Foundation for storage, but could we also add a symlink from `~/.swiftpm/repository-cache` to that directory? This makes it easier for users which may have experience with other CLI tools, especially on non-Apple platform, to find the cache directory if they need to.

There is no symlink function in TSC or am I missing something. If not I can open a quick PR for that.

tgymnich commented:
If an error occurs due to a corrupt cache or if we cannot create the cache directory for some reason, should we try recovering from that by just performing a "normal" fetch without the cache?

tgymnich commented:
> * Can we add some tests, especially around concurrent access to the cache? Anecdotally, this seems to be working fine for me locally, but we should have tests to keep that working over time.

Since we are using [an approach which keeps the per-project cache](https://github.com/apple/swift-package-manager/pull/2835#issuecomment-690771498), the only place I can imagine where something could go wrong would be right here:
```
try fileSystem.withLock(on: cachedRepositoryPath, type: .exclusive) {
   // Fetch the repository into the cache.
   try self.provider.fetch(repository: handle.repository, to: cachedRepositoryPath, update: update)
   // Copy the repository from the cache into the repository path.
   try self.provider.copy(from: cachedRepositoryPath, to: repositoryPath)
}
```
Since our `FileLock` is already unit tested I don't think it makes sense to add additional tests.

abertelrud commented:
@swift-ci please smoke test

abertelrud commented:
Thanks for the additional changes!  I'll review them later today.  I don't know the answers to the questions yet, but will address them when I take a look.

tgymnich commented:
forgot to push my last commit. Test should be fine now.

MaxDesiatov commented:
Isn't there still a conflict in `Sources/SourceControl/InMemoryGitRepository.swift`?

abertelrud commented:
Yes, I think `InMemoryGitRepository` has moved into test support now, so it looks as if there's a conflict.

Thanks @tgymnich for your continued iterations on this!  To answer a few of the questions:
- recovery from corrupt cache entries sounds great, but could as far as I am concerned by added as a follow-on PR
- there probably should be a `FileManager` function to create a symlink — I'm surprised it isn't already there; I think it would make sense to add in a separate PR that this one builds on
- to the question of TSC; I think it would make sense to have a separate PR that vendors the FileLock changes (and symlink addition if add that)

abertelrud commented on Sources/SourceControl/RepositoryManager.swift:
```diff
@@ -306,6 +319,33 @@ public class RepositoryManager {
+                }
+            } catch {
+                // Fetch without populating the cache in the case of an error.
+                print("Skipping cache due to an error: \(error)")
```
Probably need to find a way to surface this error in some other way than a print statement (for clients such as IDEs, etc).  Not sure yet if that means adding a delegate here or something like that.

tgymnich replied:
I think we can keep this as a print for now. Since this is an unimportant message, which does not affect the user directly e.g. the fetch will still work perfectly fine. It's just a hint for us. This will also appear in the build logs, right?

abertelrud replied:
I think this is fine to clear up in a future PR, in the interest of getting this integrated.  From `swift` CLI tools, it will be intermixed with the other `stdout` output... but for clients of libSwiftPM, print statements don't show up in any build logs.  Only content emitted to `DiagnosticsEngine`s do.  The output will instead appear on the `stdout` of whatever tool or app is linking against libSwiftPM.

tomerd commented:
@tgymnich this looks great, looking forward to getting it merged soon

abertelrud commented:
Thanks a lot for converting the fetch details to a struct!  I think that will be quite extensible to future information.

I, too, am really looking forward to this change!

neonichu commented:
@swift-ci please smoke test

neonichu commented:
@tgymnich Thanks for landing https://github.com/apple/swift-tools-support-core/pull/165, do you want to update this to resolve conflicts?

tgymnich commented:
Sure!

tgymnich commented:
This won't work yet because `LocalFileSystem.cachesDirectory` returns `AbsolutePath("file:///Users/tim/Library/Caches/").
Which is not allowed:
```
2020-11-16 23:37:53.227491+0100 xctest[53729:462652] Precondition failed: Failure normalizing file:///Users/tim/Library/Caches/, absolute paths should start with '/': file TSCBasic/Path.swift, line 537
```
I will try to fix this next.

tgymnich commented:
PR: https://github.com/apple/swift-tools-support-core/pull/168

tgymnich commented:
I've been thinking about ignoring local packages for caching:
```
let isLocal = (try? AbsolutePath(validating: handle.repository.url)) != nil
```
I'm not sure though if this is the best way to find out whether a package is local given a `RepositoryHandle`.

tgymnich commented:
How do we want to handle caching in our tests?
The mock workspace cache is automatically cleaned up after the tests ends either by using an `InMemoryFileSystem` or a `RerootedFileSystem`.
For integration tests where we invoke the whole swiftpm CLI we use the global cache which is not cleaned up.

- We might need to clear the cache before / after running those tests
- Ignoring local packages would prevent caching of test packages.

neonichu commented:
@swift-ci please smoke test

neonichu commented on Sources/Commands/SwiftTool.swift:
```diff
@@ -25,6 +25,7 @@ import SPMBuildCore
 import Build
 import XCBuildSupport
 import Workspace
+import Basics
```
This is a typo, right? I don't think there is a `Basics` module?

tgymnich replied:
This is not a typo. It's used for the `FileSystem` extensions which give us `var dotSwiftPM: AbsolutePath` and `var swiftPMCacheDirectory: AbsolutePath` in FileSystem+Extensions.swift.
See Package.swift:106 for the module definition.

tomerd replied:
@neonichu allot happened while you were away :D

tgymnich replied:
The smoke tests also seem to be unaware of this module.

tomerd replied:
ugh we missed adding it in the CMake setup

tomerd replied:
https://github.com/apple/swift-package-manager/pull/3064

tomerd commented:
@swift-ci please smoke test

abertelrud commented on Sources/SourceControl/Repository.swift:
```diff
@@ -73,7 +73,7 @@ public protocol RepositoryProvider {
     ///
     /// - Throws: If there is any error fetching the repository.
-    func fetch(repository: RepositorySpecifier, to path: AbsolutePath) throws
+    func fetch(repository: RepositorySpecifier, to path: AbsolutePath, update: Bool) throws
```
Since the protocol is being changed, could we add an extension here with an implementation of the new protocol method to call the old one? Otherwise clients will immediately fail to compile.  Thanks.

tgymnich replied:
Forgot about your comment from 23 days ago. I think we can go without this boolean flag.

abertelrud commented on Sources/SourceControl/Repository.swift:
```diff
@@ -118,6 +118,12 @@ public protocol RepositoryProvider {
+    /// - Parameters:
+    ///   - sourcePath: the source path.
+    ///   - destinationPath: the destination  path.
+    func copy(from sourcePath: AbsolutePath, to destinationPath: AbsolutePath) throws
```
Since the protocol is being changed, could we add an extension here with a default implementation of the new protocol method (to for example just to a regular copy in the file system)? Otherwise clients will immediately fail to compile.

tgymnich replied:
done

abertelrud replied:
Thanks!

abertelrud commented on Sources/Commands/Options.swift:
```diff
@@ -136,6 +136,12 @@ public struct SwiftToolOptions: ParsableArguments {
+    @Option(help: "Specify the global repository cache directory")
+    var cachePath: AbsolutePath?
+
+    @Flag(help: "Skip the cache when fetching")
```
Should this mention the repository cache specifically, or is this intended to skip all caches (including any ones added in the future)?

abertelrud reviewed (changes requested):
Looks good to me; a couple of comments about adding default implementations of new protocol methods, and a comment about a left-over print statement in the code.

tgymnich commented on Sources/SourceControl/GitRepository.swift:
```diff
@@ -46,18 +46,25 @@ public class GitRepositoryProvider: RepositoryProvider {
         // NOTE: We intentionally do not create a shallow clone here; the
         // expected cost of iterative updates on a full clone is less than on a
         // shallow clone.
+        if update {
```
Originally I wanted to keep this in line with `cloneCheckout` which also has a boolean flag for an assumption about the repository. Since it is quite simple to check for a pre-existing git repository in this case, I think we can just use `localFileSystem.exists(path.appending(component: ".git"))`.

abertelrud replied:
Makes sense.  I didn't notice that there was already a flag in one of the others.  Thanks for simplifying it in the latest iterations; it can always be added back if it's really needed.

abertelrud commented on Sources/SourceControl/Repository.swift:
```diff
@@ -118,12 +118,23 @@ public protocol RepositoryProvider {
     }
+
+    public func copy(from sourcePath: AbsolutePath, to destinationPath: AbsolutePath) throws {
+        fatalError("Unimplemented")
```
This is only called if caching is enabled, right?  So that existing clients would only get this fatal error if they enable caching?

tgymnich replied:
This is only called once at RepositoryManager:369 and only if a cachePath is supplied.

abertelrud replied:
Great, thanks!

tgymnich commented on Sources/SourceControl/RepositoryManager.swift:
```diff
@@ -13,14 +13,15 @@ import class Foundation.OperationQueue
 /// Delegate to notify clients about actions being performed by RepositoryManager.
 public protocol RepositoryManagerDelegate: class {
     /// Called when a repository is about to be fetched.
-    func fetchingWillBegin(handle: RepositoryManager.RepositoryHandle)
```
@abertelrud this need a default implementation too, right?

abertelrud replied:
Technically, all of the methods in the delegate protocol are optional, so they should have default implementations.  But since `fetchWillBegin(handle:)` is an existing method that any delegate must already be implementing, I don't think it's needed for this PR.

Before tagging a semver 1.0 of libSwiftPM we should revisit and standardize all of these, but since the method preexisted this PR I don't think anything else is needed here.

tgymnich commented on Sources/Workspace/Workspace.swift:
```diff
@@ -45,10 +45,10 @@ public protocol WorkspaceDelegate: class {
 
     /// The workspace has started fetching this repository.
-    func fetchingWillBegin(repository: String)
+    func fetchingWillBegin(repository: String, fetchDetails: RepositoryManager.FetchDetails?)
```
same here

abertelrud replied:
Yes, the general approach I'd recommend (to avoid breaking clients) when adding new delegate callbacks is to have the code call the new method, and then provide a default implementation of the new method that calls the old one.

Something like:
```
    public func fetchingWillBegin(repository url: String, fetchDetails: RepositoryManager.FetchDetails?) {
        self.fetchingWillBegin(repository: url)
    }
```

The clients can update at their leisure.  Marking the old callback as deprecated would be fine, I think.

abertelrud replied:
Once there are semantic versions of libSwiftPM this won't be an issue (we'd just bump the semantic version when making breaking changes), but while all the clients still have to have branch dependencies, it's difficult to stage the adoption without trying to coordinate commits across repositories.

neonichu commented:
@swift-ci please smoke test

abertelrud commented on Sources/SourceControl/RepositoryManager.swift:
```diff
@@ -30,8 +31,8 @@ public protocol RepositoryManagerDelegate: class {
 public extension RepositoryManagerDelegate {
-    func fetchingWillBegin(handle: RepositoryManager.RepositoryHandle) {}
-    func fetchingDidFinish(handle: RepositoryManager.RepositoryHandle, error: Swift.Error?) {}
+    func fetchingWillBegin(handle: RepositoryManager.RepositoryHandle, fetchDetails: RepositoryManager.FetchDetails?) {}
```
Instead of removing `func fetchingWillBegin(handle: RepositoryManager.RepositoryHandle) {}` altogether, I'd mark it deprecated and implement `func fetchingWillBegin(handle:fetchDetails:)` to call it in the extension.  This lets clients that haven't yet implemented it keep working.

This will all be made easier once there is a semantically versioned API for libSwiftPM.  Right now it's a hassle but changing the API without staging causes a lot of downstream problems for any clients.

neonichu commented:
@swift-ci please smoke test

neonichu commented:
Looks like there's still an issue with the `Basics` module on CI, will check locally.

neonichu commented:
Ah, the issue is a missing dependency in the CMake configuration. Since `SourceControl` now depends on `Basics`, a dependency needs to be added:

```
% git diff Sources/SourceControl/CMakeLists.txt 
diff --git a/Sources/SourceControl/CMakeLists.txt b/Sources/SourceControl/CMakeLists.txt
index 5c45fe50..58c9c9b1 100644
--- a/Sources/SourceControl/CMakeLists.txt
+++ b/Sources/SourceControl/CMakeLists.txt
@@ -12,6 +12,7 @@ add_library(SourceControl
   RepositoryManager.swift)
 
 target_link_libraries(SourceControl PUBLIC
+  Basics
   TSCBasic
   TSCUtility)
 
```

tgymnich commented:
Done in #3070

tomerd commented:
@swift-ci please smoke tes

tomerd commented:
@swift-ci please smoke test

abertelrud commented on Sources/SourceControl/RepositoryManager.swift:
```diff
@@ -30,6 +39,17 @@ public protocol RepositoryManagerDelegate: class {
+
+    @available(*, deprecated)
+    func fetchingDidFinish(handle: RepositoryManager.RepositoryHandle, fetchDetails: RepositoryManager.FetchDetails?, error: Swift.Error?) {
+        fetchingDidFinish(handle: handle, error: error)
```
Great, thank you, this looks good!

abertelrud reviewed (approved):
Looks good to me.  Thanks!

tomerd commented:
amazing work @tgymnich  🙇‍♂️ ❤️ 

can't wait to get this merged and use it in our day-to-day workflows 🥳

tgymnich commented:
You are welcome 😊. Thank you all for your reviews and helpful comments.

tgymnich commented:
Determining if a git repository exists by checking if `".git"` exists does't work when we don't do a checkout. Now we are using  `git rev-parse --is-inside-git-dir --is-inside-work-tree` which works in both cases.

MaxDesiatov commented:
@swift-ci please smoke test

tgymnich commented:
> The tests are also using the local cache. We might need another change to swift-tools-support to modify `SwiftPMProduct.SwiftPackage.execute(_ args: [String], packagePath: AbsolutePath? = nil, env: [String : String]? = nil))` maybe adding a `--skip-cache` flag by default.
> 
> I should also look into exclusively caching remote dependencies and excluding local dependencies.

I went ahead and added some integration tests for caching in `PackageToolTests.swift` and some Unit tests in `RepositoryManagerTests.swift`. I added a boolean flag and a environment variable to disable skipping of local packages.

MaxDesiatov commented:
@swift-ci please smoke test

abertelrud commented:
> I went ahead and added some integration tests for caching in `PackageToolTests.swift` and some Unit tests in `RepositoryManagerTests.swift`. I also added a boolean flag and an environment variable to disable skipping of local packages when running tests.

Thanks for the additional changes here, @tgymnich, and apologies for the slow response.  I see the intent with these flags, but it does make it a bit more difficult for libSwiftPM clients (such as IDEs) to enable this feature, since the cache is off by default.  I would suggest not framing the boolean in terms of running unit tests, but rather that there is a parameter that controls whether or not to use a cache, and then both tests and actual clients of libSwiftPM can set that parameter accordingly.  The command line argument in the swift subcommand would then either set that parameter or not depending on the flag that's passed.  In other words, I don't think that the control over whether to use the cache is a test-driven feature, but rather a general switch of the infrastructure.

It seems to me that since the cache directory is optional and is passed in as a parameter already, that is the only parameter that's really needed.  If the client passes in a repository cache, then it's used — otherwise it is not.  And then the command line tool that decides whether to pass the directory or nil does so by looking at the command line flag, and an IDE could have other criteria for whether or not to pass a cache directory.  Does that seem reasonable?

abertelrud commented:
> It seems to me that since the cache directory is optional and is passed in as a parameter already, that is the only parameter that's really needed. If the client passes in a repository cache, then it's used — otherwise it is not. And then the command line tool that decides whether to pass the directory or nil does so by looking at the command line flag, and an IDE could have other criteria for whether or not to pass a cache directory. Does that seem reasonable?

Actually:  it looks as if the line:
```
let cachePath = !options.skipCache ? options.cachePath ?? defaultCachePath : nil
```
in SwiftPackageTool does this already.  So I think the only thing that needs to happen is that the test code (e.g. in MockWorkspace) would check the user default in order to decide whether or to pass a cache path or nil, just as any other client would.

abertelrud commented:
> I went ahead and added some integration tests for caching in `PackageToolTests.swift` and some Unit tests in `RepositoryManagerTests.swift`. I also added a boolean flag and an environment variable to disable skipping of local packages when running tests.

Thanks for the additional changes, @tgymnich.  For the controlling whether local packages are cached, it might be better to use a more general parameter than something specifically intended for unit tests embedded in the general code.  If the parameter has a sensible default value, then it could be left to the unit test to look at the environment variable to decide whether to enable the behavior.  Just a thought, but that would make the general code a bit cleaner.

tgymnich commented:
> Thanks for the additional changes, @tgymnich. For the controlling whether local packages are cached, it might be better to use a more general parameter than something specifically intended for unit tests embedded in the general code. If the parameter has a sensible default value, then it could be left to the unit test to look at the environment variable to decide whether to enable the behavior. Just a thought, but that would make the general code a bit cleaner.

I think caching of local packages makes only sense in the context of unit tests. I see only two ways of enabling this behavior from within `PackageToolTests`: a command line argument or an environment variable.

abertelrud commented:
> I think caching of local packages makes only sense in the context of unit tests. I see only two ways of enabling this behavior from within `PackageToolTests`: a command line argument or an environment variable.

Right, I agree that it doesn't make sense as an external-facing feature, I was just talking about how the main code paths check for this condition.  We can always refine this in a separate PR, though.

I'll take another quick look but as far as I am concerned this is ready to merge.  Any objections?  Thanks for all your work on this!  Can't wait to use this!

tgymnich commented:
👍🏼 let's refine this later. For me this is ready to be merged as well.

abertelrud commented:
Fantastic.  Thanks for your work on this!

tomerd commented:
this is fantastic 🥳 ❤️

neonichu commented:
This is great 🎉 Thank you @tgymnich!
